profiling-ffi/src/string_storage.rs

```diff
@@ -0,0 +1,209 @@
+// Copyright 2023-Present Datadog, Inc. https://www.datadoghq.com/
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::Context;
+use datadog_profiling::collections::string_storage::ManagedStringStorage as InternalManagedStringStorage;
+use ddcommon_ffi::{CharSlice, Error, MaybeError, StringWrapper};
+use libc::c_void;
+use std::{ffi::CStr, rc::Rc, sync::RwLock};
+
+#[repr(C)]
+pub struct ManagedStringId {
+    pub value: u32,
+}
+
+#[repr(C)]
+pub struct ManagedStringStorage {
+    // This may be null, but if not it will point to a valid InternalManagedStringStorage.
+    inner: *const c_void, /* Actually *RwLock<InternalManagedStringStorage> but cbindgen doesn't
+                           * opaque RwLock */
+}
+
+#[allow(dead_code)]
+#[repr(C)]
+pub enum ManagedStringStorageNewResult {
+    Ok(ManagedStringStorage),
+    #[allow(dead_code)]
+    Err(Error),
+}
+
+#[no_mangle]
+#[must_use]
+pub unsafe extern "C" fn ddog_prof_ManagedStringStorage_new() -> ManagedStringStorageNewResult {
+    let storage = InternalManagedStringStorage::new();
+
+    ManagedStringStorageNewResult::Ok(ManagedStringStorage {
+        inner: Rc::into_raw(Rc::new(RwLock::new(storage))) as *const c_void,
+    })
+}
+
+#[no_mangle]
+/// TODO: @ivoanjo Should this take a `*mut ManagedStringStorage` like Profile APIs do?
```

Comment:
Ugh, this is something even in C I go back and forth on. It's one of those "do I trust the user or do I be more defensive?" things. Setting the C pointer to null makes it easier to debug when things go wrong, and can sometimes even prevent further things from going wrong. But it's theoretically wholly wasted work because nobody should use the thing after it's been dropped...

Reply:
I still don't have the "answer" to this one. All I can say is that I made the others re-assign the pointer null on drop when it was feasible, so I clearly thought at the time that it was a good idea.